File: string_matching/kmp.zig
const std = @import("std");

pub fn search(pattern: []const u8, text: []const u8) ?usize {
    if (pattern.len == 0) return 0;
    var allocator = std.heap.page_allocator;
    var lps = allocator.alloc(usize, pattern.len) catch return null;
    defer allocator.free(lps);

    var len: usize = 0;
    lps[0] = 0;
    var i: usize = 1;
    while (i < pattern.len) {
        if (pattern[i] == pattern[len]) {
            len += 1;
            lps[i] = len;
            i += 1;
        } else if (len != 0) {
            len = lps[len - 1];
        } else {
            lps[i] = 0;
            i += 1;
        }
    }

    var ti: usize = 0;
    var pj: usize = 0;
    while (ti < text.len) {
        if (pattern[pj] == text[ti]) {
            ti += 1;
            pj += 1;
            if (pj == pattern.len) return ti - pj;
        } else if (pj != 0) {
            pj = lps[pj - 1];
        } else {
            ti += 1;
        }
    }
    return null;
}

File: string_matching/kmp.c
#include "kmp_c.h"
#include <stdlib.h>
#include <string.h>

static void build_lps(const char *pattern, int *lps, size_t m) {
    size_t len = 0;
    lps[0] = 0;
    for (size_t i = 1; i < m;) {
        if (pattern[i] == pattern[len]) {
            lps[i++] = ++len;
        } else if (len != 0) {
            len = lps[len - 1];
        } else {
            lps[i++] = 0;
        }
    }
}

int kmp_search(const char *pattern, const char *text) {
    size_t m = strlen(pattern);
    size_t n = strlen(text);
    if (m == 0) return 0;
    int *lps = (int*)malloc(sizeof(int) * m);
    if (!lps) return -1;
    build_lps(pattern, lps, m);
    size_t i = 0, j = 0;
    while (i < n) {
        if (pattern[j] == text[i]) {
            ++i; ++j;
            if (j == m) {
                free(lps);
                return (int)(i - j);
            }
        } else if (j != 0) {
            j = lps[j - 1];
        } else {
            ++i;
        }
    }
    free(lps);
    return -1;
}

